doc/connectivity-semantics-and-api.md

```diff
@@ -28,7 +28,10 @@ through TLS handshake (or equivalent) and all subsequent attempt to communicate
 have succeeded (or are pending without any known failure ).
 
 TRANSIENT_FAILURE: There has been some transient failure (such as a TCP 3-way
-handshake timing out or a socket error). Channels in this state will eventually
+handshake timing out or a socket error). New RPCs should, by default, fail
```

Comment:
I'm a little leery of explicitly tying RPC handling to channel state, because the channel state notifications are asynchronous with respect to the RPC handling.  In other words, at the point at which we're trying to decide which backend to send an RPC to, I think we should actually try to send it and see what happens instead of checking the most recent connectivity state notification, because it could be the case that the most recent notification was for TRANSIENT_FAILURE but the channel has since recovered and is now in state READY, only we haven't yet received the notification about that.

I think it would be better to consider the connectivity state to be a parallel but independent path to the RPC handling path.  In other words, the same conditions that cause us to go into state TRANSIENT_FAILURE should cause new RPCs to be failed with UNAVAILABLE by default.  But that's different from saying that we should check the state itself in order to make this decision.

Reply:
> I'm a little leery of explicitly tying RPC handling to channel state, because the channel state notifications are asynchronous with respect to the RPC handling.

Them being asynchronous is the exact reason I'm okay with this. The channel state API is guaranteed to be imprecise, so we still have leeway as we just have to produce a _plausible_ ordering.

> I think we should actually try to send it and see what happens instead of checking the most recent connectivity state notification

Oh, definitely. I don't see this as disallowing that.

> only we haven't yet received the notification about that.

Notifications, by their very nature are after-the-fact. Even in a system where they were immediately delivered there would still be a delay for application processing, so it'd _still_ be after-the-fact. The behavior here is instantaneous, which the user never gets to see.

> I think it would be better to consider the connectivity state to be a parallel but independent path to the RPC handling path.

:-/ This is complex enough already. I see that as making it even worse. I like this definition because it is concise, non-repetitive, and is comprehensible to users.

Reply:
I see your point, but I think we can improve the wording here without adding too much complexity.  For example, how about changing "New RPCs" to "New RPCs that encounter these errors"?  That way, we're making it clear that this state indicates these kinds of failures and that these failures cause certain behavior for RPCs, without actually explicitly tying the RPC behavior to the state.

Reply:
I'm not really sure what "these errors" means "New RPCs that encounter these errors". I've changed the language to be less of an implementation specification to be more "this is the sort of behavior you will see." I doubt it will fully satisfy you, but maybe it gets closer.

Reply:
The phrase "these errors" was meant to refer to the text "such as a TCP 3-way handshake timing out or a socket error" from the previous sentence.  If that's not clear enough, how about something like the following:

There has been some transient failure, such as a failed attempt to resolve the target name, a failed attempt to connect to a server, or an established connection becoming disconnected.  New RPCs that encounter failures to establish a connection to a server will fail immediately with status UNAVAILABLE, but implementations may support options like [wait-for-ready](wait-for-ready.md) that modify that behavior.